mackerel-plugin-multicore/multicore.go

```diff
@@ -0,0 +1,402 @@
+}
+
+type cpuPercentages struct {
+	User    float64
+	Nice    float64
+	System  float64
+	Idle    float64
+	IoWait  float64
+	Irq     float64
+	SoftIrq float64
+	Steal   float64
+	Guest   float64
+}
+
+func getProcStat() (string, error) {
+	contentbytes, err := ioutil.ReadFile("/proc/stat")
+	if err != nil {
+		return "", err
+	}
+	return string(contentbytes), nil
+}
+
+func parseFloats(values []string) ([]float64, error) {
+	var result []float64
+	for _, v := range values {
+		f, err := strconv.ParseFloat(v, 64)
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, f)
+	}
+	return result, nil
+}
+
+func fill(arr []float64, elementCount int) []float64 {
+	arraySize := len(arr)
+	if arraySize < elementCount {
+		for i := arraySize; i < elementCount; i++ {
+			arr = append(arr, 0.0)
+		}
+	}
+	return arr
+}
+
+func parseProcStat(str string) (map[string]*procStats, error) {
+	var result = make(map[string]*procStats)
+	for _, line := range strings.Split(str, "\n") {
+		if cpuNumberPattern.MatchString(line) {
+			var key = strings.Fields(line)[0]
+			var values = strings.Fields(line)[1:]
+
+			floatValues, err := parseFloats(values)
+			if err != nil {
+				return nil, err
+			}
+			filledValues := fill(floatValues, 9)
+
+			total := 0.0
+			for _, v := range floatValues {
+				total += v
+			}
+
+			ps := &procStats{
+				User:    filledValues[0],
+				Nice:    filledValues[1],
+				System:  filledValues[2],
+				Idle:    filledValues[3],
+				IoWait:  filledValues[4],
+				Irq:     filledValues[5],
+				SoftIrq: filledValues[6],
+				Steal:   filledValues[7],
+				Guest:   filledValues[8],
+				Total:   total,
+			}
+			result[key] = ps
+
+		} else {
+			break
+		}
+	}
+	return result, nil
+}
+
+func collectProcStatValues() (map[string]*procStats, error) {
+	procStats, err := getProcStat()
+	if err != nil {
+		return nil, err
+	}
+	return parseProcStat(procStats)
+}
+
+func saveValues(tempFileName string, values map[string]*procStats, now time.Time) error {
+	f, err := os.Create(tempFileName)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	s := saveItem{
+		LastTime:       time.Now(),
+		ProcStatsByCPU: values,
+	}
+
+	encoder := json.NewEncoder(f)
+	err = encoder.Encode(s)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func fetchLastValues(tempFileName string) (map[string]*procStats, time.Time, error) {
+	f, err := os.Open(tempFileName)
+	if err != nil {
+		return nil, time.Now(), err
+	}
+	defer f.Close()
+
+	var stat saveItem
+	decoder := json.NewDecoder(f)
+	err = decoder.Decode(&stat)
+	if err != nil {
+		return stat.ProcStatsByCPU, stat.LastTime, err
+	}
+	return stat.ProcStatsByCPU, stat.LastTime, nil
+}
+
+func calcCPUUsage(currentValues map[string]*procStats, now time.Time, lastValues map[string]*procStats, lastTime time.Time) (map[string]*cpuPercentages, error) {
+
+	result := make(map[string]*cpuPercentages)
+	for key, current := range currentValues {
+		last := lastValues[key]
```

Comment:
I'd be better you ensure this value is not nil but it'll work well without the check in actual situation...